### PR TITLE
Preserve LATERAL UNNEST during SqlNode validation

### PR DESCRIPTION
**Pull request by KevinGe00**
**Summary**

This is a follow up to #98, where LATERALs were only preserved during validation if they appeared in the form `... LATERAL (SELECT ....`. This PR induces the preservation behavior also in the `.. LATERAL (UNNEST ...` case, this is needed for Coral to produce correct translations on views (see [CoralSqlNodeToSparkSqlNodeConverter](https://github.com/linkedin/coral/blob/master/coral-spark/src/main/java/com/linkedin/coral/spark/CoralSqlNodeToSparkSqlNodeConverter.java) for details).

**Testing**
- new unit test to test that this only a preservation mechanism, we never add LATERALs that were never there to begin with
   - note that we cannot write a LATERAL UNNEST test case here in Calcite since it will not parse, see [here](https://github.com/linkedin/linkedin-calcite/blob/c10eb657eeb72673ad2d2ebe6ed959ce6aaee555/core/src/test/java/org/apache/calcite/sql/parser/SqlParserTest.java#L7405). Instead we will use an acompanying Coral unit test that will generate the LATERAL UNNEST sqlnode case we need to test
- new unit test in coral with these changes: https://github.com/linkedin/coral/pull/542
- mvn clean build (also runs all unit tests) `./mvnw clean install -Dgpg.skip -f core/pom.xml`
- clean build in coral with these changes `./gradlew clean build`
- correct translation with these changes for production view text which was previously incorrect 
- regression testing on all production views [pending]

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java
Comment (KevinGe00) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java:3144]: This was incorrect, we want to recurse into same function (`validateFrom`) and not into `validateQuery`.

File: core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java
Comment (KevinGe00) [inline core/src/main/java/org/apache/calcite/sql2rel/SqlToRelConverter.java:2004]: This is needed to work correctly with Coral's [HiveSqlToRelConverter](https://github.com/linkedin/coral/blob/c51d3b5b457e8e4af17d719cf85b79220c3aff12/coral-hive/src/main/java/com/linkedin/coral/hive/hive2rel/HiveSqlToRelConverter.java#L76) class which extends this class and overrides this function.

**Comment by aastha25** [comment]
@KevinGe00 in general (1) lets add the commands used to build & test, especially how to i-test with Coral, in the PR description. it helps build confidence in the changes and it'll be useful for future work as well. (2) Please file the Coral PR and cross reference it here so that reviewers can follow testing done related to `new unit test in coral with these changes`

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java
Comment (aastha25) [inline core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: should this be `if (lateral && (kind == SqlKind.SELECT || kind == SqlKind.UNNEST))` to denote that LATERAL must be present? 
otherwise this could match `SELECT ... FROM (UNNEST ...) AS t` as well

File: core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java
Comment (aastha25) [inline core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java:11541]: not sure I agree. Our spec for Coral IR sqlNode suggests 'LATERAL' will be 'preserved'. If LATERAL is not available in the input, we shouldnt add it. There should be an explicit transformer which introduces it.

File: core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java
Comment (KevinGe00) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: In this function (and Calcite in general), UNNEST will always imply LATERAL (i.e `lateral = true`), see [here](https://github.com/linkedin/linkedin-calcite/blob/li-1.21.0/core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java#L2273). So in theory we should never see `SELECT ... FROM (UNNEST ...) AS t` but I agree that the conditional doesn't read as nice so changed :)

Comment (KevinGe00) [reply core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java]: In this function (and Calcite in general), UNNEST will always imply LATERAL (i.e lateral = true), see [here](https://github.com/linkedin/linkedin-calcite/blob/li-1.21.0/core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java#L2273), so both conditionals are equivalent but I agree that this conditional doesn't read as nice so changed :)

File: core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java
Comment (KevinGe00) [reply core/src/test/java/org/apache/calcite/test/SqlValidatorTest.java:11541]: fixed!

**Comment by KevinGe00** [comment]
> @KevinGe00 in general (1) lets add the commands used to build & test, especially how to i-test with Coral, in the PR description. it helps build confidence in the changes and it'll be useful for future work as well. (2) Please file the Coral PR and cross reference it here so that reviewers can follow testing done related to `new unit test in coral with these changes`

i-testing with Coral can only be done after merging and publishing these changes, then using the newest Calcite version containing these changes in a Coral feature branch in the regression pipeline, same thing done for #98.

**Comment by aastha25** [comment]
Approving to unblock i-testing. Let's revisit the code changes of this PR during / after the i-testing.

**Review by aastha25 — APPROVED** [review]

**Pull request merged**
